R/utils.R

```diff
@@ -14,6 +14,25 @@ jackknife = function(x, theta) {
   jack.se
 }
 
+#' Bootstrap standard error of function `theta` at samples `n`.
+#' @param B repetitions
+#' @param n number of samples
+#' @param k number of draws
+#' @param estimate estimate
+#' @param theta a function which returns a scalar estimate
+#' @keywords internal
+bootstrap = function(B, n, k, estimate, theta, replace = TRUE) {
+  u = rep(0, B)
+  for (i in 1:B) {
+    ind = sample(1:n, k, replace = replace)
+    u[i] = theta(ind)
+  }
+  boot.se = sqrt(mean((c(estimate) - u)^2, na.rm = TRUE))
+  # boot.se=sqrt(var(u, na.rm = TRUE))
+
+  boot.se
+}
+
```

Comment:
This is a pretty confusing interface. Maybe replace it with two functions.
1. bootstrap(x,theta,replications) where x and theta match what is used for the jackknife and we always sample a list of the same length as x with replacement
2. subsample(x,theta,size,replications) where we subsample without replacement

Reply:
Sorry, I should have given you the implementations. These functions are essentially standard and encode invariants useful when thinking about the method. Please use the following.

bootstrap = function(x,theta,replications) {
    replicate(replications, theta(sample(x, replace=TRUE)))
}
subsample = function(x,theta,size,replications) {
   replicate(replications, theta(sample(x, size, replace=FALSE)))
}

But we're not actually subsampling, so we don't need the second. Sorry about that.
In terms of these, we can implement standard errors as follows:

bootstrap_se = function(estimate, replications=200) { 
   theta = function(ind) { synthdid_estimate(Y=setup$Y[sort(ind)], N0=sum(ind <= setup$N0), T0=setup$T0, X=setup$X[sort(ind), ,], <options> }
sd(bootstrap(1:nrow(Y),  theta, replications=replications))
}

placebo_se = function(estimate) {
  N1=nrow(setup$Y)-setup$N0
  theta = function(ind) { synthdid_estimate(Y=setup$Y[ind], N0=length(ind)-N1,  T0=setup$T0,  X=setup$X[ind, ,], <options> }
  sd(replicate(replications, theta(sample(1:setup$N0))))
}

Reply:
Ok, it is updated

Reply:
The <options> should be the same ones passed to produce the original estimate, so the optimizer gets the same stopping parameters, we use sc or did if that is what we used in the first place, etc. 

And we should return NA when there is one treated unit for the bootstrap and jackknife, but not the placebo --- that's what the placebo estimate is for.